lib/logger.js

```diff
@@ -107,27 +107,64 @@ function createLogger (options) {
  * of a Fastify compatible logger.
  *
  * @param {object} logger Object to validate.
+ * @param {boolean?} strict `true` if the object must be a logger (always throw if any methods missing)
  *
  * @returns {boolean} `true` when the logger meets the requirements.
  *
  * @throws {FST_ERR_LOG_INVALID_LOGGER} When the logger object is
  * missing required methods.
  */
-function validateLogger (logger) {
+function validateLogger (logger, strict) {
   const methods = ['info', 'error', 'debug', 'fatal', 'warn', 'trace', 'child']
-  const missingMethods = methods.filter(method => !logger[method] || typeof logger[method] !== 'function')
+  const missingMethods = logger
+    ? methods.filter(method => !logger[method] || typeof logger[method] !== 'function')
+    : methods
 
   if (!missingMethods.length) {
     return true
-  } else if (missingMethods.length === methods.length) {
+  } else if ((missingMethods.length === methods.length) && !strict) {
     return false
   } else {
     throw FST_ERR_LOG_INVALID_LOGGER(missingMethods.join(','))
   }
 }
 
+/**
+ * Utility for creating a child logger with the appropriate bindings, logger factory
+ * and validation.
+ * @param {import('../fastify').FastifyBaseLogger} logger
+ * @param {import('../fastify').RawRequestDefaultExpression<any>} req
+ * @param {string} reqId
+ * @param {import('../fastify').FastifyServerOptions} options
+ * @param {import('../types/logger.js').ChildLoggerOptions?} loggerOpts
+ */
+function createChildLogger (logger, req, reqId, options, loggerOpts) {
+  const loggerBindings = {
+    [options.requestIdLogLabel]: reqId
+  }
+  const child = options.childLoggerFactory(logger, loggerBindings, loggerOpts || {}, req)
```

Comment:
Wondering if the `childLoggerFactory` should have access to the `this` context linked to the `fastify` instance 🤔

Reply:
I was thinking so too, however I just haven't understood the code enough to know how to reference the fastify instance in these 3 files. 😓 

Furthermore with the typings, to strictly/properly specify the 5 generic parameters of the `FastifyInstance` type, I'd need to add `RawRequest`, `RawReply`, `TypeProvider` generic params to the `FastifyServerOptions` type which may be a breaking change? Otherwise we leave those generic parameters at their default value which would also work, although not type-correct in the strictest sense.

Reply:
> Wondering if the childLoggerFactory should have access to the this context linked to the fastify instance

Ok I bound `this` to the Fastify instance. In some cases (404) is is the root instance. In normal cases it is the encapsulated instance (I think).

Really need an export to check this last commit very carefully because I'm not 100% confident that I've referenced the correct Fastify instance in each case.

Reply:
@mcollina did you carefully check that I am referencing the correct fastify instance in each case?

Reply:
>Really need an export to check this last commit very carefully because I'm not 100% confident that I've referenced the correct Fastify instance in each case.

See my previous comment, you did it right, but better to fallback into the instance owner of the `forOhFour` 🙂